File: leetcode/java/0124. Binary Tree Maximum Path Sum.java
package leetcode.java;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// Time complexity:O(N), where N is the number of nodes
// Space complexity:O(N)

class Solution {

    private int maxSum = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        this.dfs(root);
        return maxSum;
    }

    public int dfs(TreeNode source) {

        if (source == null) {
            return 0;
        }

        int maxLeftNoSplit = Integer.max(this.dfs(source.left), 0);
        int maxRightNoSplit = Integer.max(this.dfs(source.right), 0);

        this.maxSum = Integer.max(this.maxSum, source.val + maxLeftNoSplit + maxRightNoSplit);

        return source.val + Integer.max(maxLeftNoSplit, maxRightNoSplit);
    }
}

File: leetcode/python/0124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Time complexity:O(N), where N is the number of nodes
# Space complexity:O(N)

class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        maxSum = [-99999]
        self.dfs(root, maxSum)
        return maxSum[0]

    def dfs(self, source, maxSum):

        if source is None:
            return 0
        
        maxLeft = max(self.dfs(source.left), 0)
        maxRight = max(self.dfs(source.right), 0)

        maxSum[0] = max(maxSum[0], maxLeft+maxRight+source.val)

        return source.val + max(maxLeft, maxRight)